scripts/coccinelle/boolean-expression.cocci

```diff
@@ -0,0 +1,357 @@
+/// Make statements evaluate Boolean expressions
+///
+// Confidence: Low
+// Copyright (c) 2018 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+virtual patch
+
+@depends on !(file in "ext")@
+expression *X;
+statement S;
+@@
+
+- if (X)
++ if (X != NULL)
+  S
+
+@depends on !(file in "ext")@
+expression *X;
+statement S1, S2;
+@@
+
+- if (X) {
++ if (X != NULL) {
+S1
+} else {
+S2
+}
+
+@depends on !(file in "ext")@
+expression *X;
+statement S;
+@@
+
+- if (X) {
++ if (X != NULL) {
+S
+}
+
+@depends on !(file in "ext")@
+typedef uint8_t;
+typedef u8_t;
+typedef uint16_t;
+typedef u16_t;
+typedef uint32_t;
+typedef u32_t;
+{uint8_t,u8_t,uint16_t,u16_t,uint32_t,u32_t,unsigned int,int,size_t,unsigned} I;
+statement S;
+@@
+
+- if (!I)
++ if (I == 0)
+S
+
+@depends on !(file in "ext")@
+expression *E;
+@@
+
+  E ==
+- 0
++ NULL
+
+@depends on !(file in "ext")@
+expression *E;
+@@
+
+- 0
++ NULL
+  == E
+
+@depends on !(file in "ext")@
+expression *E;
+@@
+
+  E !=
+- 0
++ NULL
+
+@depends on !(file in "ext")@
+expression *E;
+@@
+
+- 0
++ NULL
+  != E
+
+
+@depends on !(file in "ext")@
+idexpression *X;
+expression E;
+statement S;
+@@
+
+- if (X && E)
++ if ((X != NULL) && E)
```

Comment:
Use `...` instead as your case won't catch:
```
int main (void)
{
	unsigned int *foo; 
	if (foo && baz && boat)
		bar++;
}
```

Reply:
use `...` instead of ? the expression `E` ?

Reply:
yeah, there are several cases, that is why I put low confidence. Also I got some problems of comparing non-pointers with NULL. Last but no least, some problems with adding `{` in cases like:
```
else if (....)  {
}
```
becoming
```
else { if (...) {
         }
}
```